File: archive/2024-06-01/string/T5_repeatedSubstringPattern.java
package org.review2024summer.string;

public class T5_repeatedSubstringPattern {
    public boolean repeatedSubstringPattern(String s) {
        char[] patt = s.toCharArray();
        int[] next = new int[patt.length];

        int j = 0;
        for (int i = 1; i < patt.length; i++) {
            while (patt[i] != patt[j] && j > 0)
                j = next[j - 1];
            if (patt[i] == patt[j])
                j++;
            next[i] = j;
        }

        int len = patt.length;
        int subLen = len - next[len - 1];
        return subLen != len && subLen != 0 && len % subLen == 0;
    }
}

File: archive/2026-04-05/ch04_string/t5_repeated_substring_pattern.py
"""LeetCode 459. Repeated Substring Pattern
https://leetcode.cn/problems/repeated-substring-pattern/
"""


class Solution:
    def repeatedSubstringPattern(self, s: str) -> bool:
        next = [0] * len(s)

        j = 0
        for i in range(1, len(s)):
            while j > 0 and s[i] != s[j]:
                j = next[j - 1]
            if s[i] == s[j]:
                j += 1
            next[i] = j

        subLen = len(s) - next[-1]
        return subLen != len(s) and subLen != 0 and len(s) % subLen == 0


if __name__ == "__main__":
    solution = Solution()
    assert solution.repeatedSubstringPattern("abab") is True
    assert solution.repeatedSubstringPattern("aba") is False
    assert solution.repeatedSubstringPattern("abcabcabcabc") is True
    print("All tests passed.")
